### ISSUE TITLE
Assigning array to span in ES.42

### ISSUE THREAD

**Issue by jack17529** [isocpp/CppCoreGuidelines#1306]
In the first f2 function instead of `span<int> a = {arr, pos};`  shouldn't we should use `gsl::span<int> a(arr.data(),pos);`

**Comment by cubbimew** [comment]
Although it's not an assignment, it's true that neither `std::span` nor `gsl::span` has a constructor that takes an `std::array` and a size (or plain array and size)

Based on the discussion in https://github.com/Microsoft/GSL/issues/442 , perhaps it is fair to remove the usage here.

**Comment by jack17529** [comment]
@cubbimew so, should I submit a PR?

**Comment by annagrin** [comment]
I understand that the constructor was removed as unnecessary by the iso committee decision. Should we re-examine this decision as well?

Thanks,
Anna

**Comment by hsutter** [comment]
Yes, we'd like to reopen this issue for the GSL implementation, and also raise it again in WG21.

**Comment by hsutter** [comment]
Editors call: For now, we've updated the example from

        span<int> a = {arr, pos};

to

        span<int> a = {arr.data(), pos};

However, this loses safety by explicitly throwing away bounds information -- calling the `span` constructor with `{arr.data(), pos}` no longer gives the constructor an opportunity to ensure that `pos` is in bounds in a checking implementation. This is a correctness regression.

Furthermore, in the case when a `span` is used as a parameter type, which we expect to be a frequent use case and was a principal design target for `span`, it means that all such call sites will be littered with `.data()` which explicitly throws away bounds information and makes all those call sites bounds-unsafe by construction. Preventing this class error was exactly the main point of adding `span` as a vocabulary type in the first place.

Upon further discussion, we think the right plan forward is to coordinate this with the newly-merged Ranges proposal, on several fronts:

1. Make sure that `std::span` (and `gsl::span`) is-a contiguous range in the Ranges concepts.
2. Solidify advice on how contiguous containers should interact with Ranges and `span`. In particular, should `span` have a constructor taking any other ContiguousRange or ContiguousContainer so as not to lose bounds information, the result being a `span` to the whole container's contents. Then those constructors should have overloads that take a second optional length parameter to refer to only the first `pos` elements (which can be checked against the actual size to ensure it is a subset which is safe).

If that direction works out, then the original code of

    span<int> s = {arr, pos};

would work for `array`, but additionally work for `vector` (which is even more common), without throwing away bounds information and therefore bounds-checkable where desired.

More importantly, passing the first `pos` elements of `arr` to a function `f(span<int>)` would be:

    f({arr,pos});

which has always been a key design goal, to avoid the bug farm of `f(arr.data(), pos)` C-style APIs.

**Comment by CaseyCarter** [comment]
Why not `span s = span{arr}.first(pos);`?

**Comment by hsutter** [comment]
@CaseyCarter Yes, I mentioned that one on the call, and that's how I started to write up the above summary.

The trouble with that one is that it works fine, and is nice and general, for writing the original 'ideal' of

    span<int>s = {arr, pos};

instead as

    span<int> s = span<int>(arr).first(pos);

    // or spelled like this, using our modern auto/CTAD conveniences:
    auto s = span(arr).first(pos);

A big concern about that is that the bread-and-butter case is actually function parameters, and replacing brittle bounds-unsafe C (ptr,len) APIs with bounds-safe (span) APIs. For that, our ideal had been to replace this C style

    // to call void f(int* data, int length);
    
    f(arr.data(), arr.size());   // pass the whole array: brittle, easy to err
    f(arr.data(), pos);   // pass the first pos elements: brittle, easy to err, and not checkable

with this

    // to call void f(span<int>);
    
    f(arr);   // pass the whole array: correct by construction
    f({arr, pos});   // pass the first pos elements: easy for span ctor to check pos <= arr.size()

With the above, I think that would be

    // to call void f(span<int>);
    
    f(arr);   // pass the whole array: ok, as above
    f(span(arr).first(pos));   // pass the first pos elements: lose convenience parity with (ptr,len) API

i.e., we lose the ability to omit `span` and get fairly natural function call syntax, including upgrading existing call sites that today are passing (ptr,len) to pass ({ptr,len}).

It works, but a primary goal is to compete with the convenience of the unsafe C API. We don't want to leave "it's more convenient to pass (ptr,len) and I know what I'm doing" rationalization to write unsafe brittle code.

**Comment by CaseyCarter** [comment]
Well, I suppose I won't suggest `f(span{arr} | view::take(pos))`, in that case ;)

**Comment by annagrin** [comment]
@CaseyCarter you can do that? ;)

@hsutter so should we add the constructor back in GSL? I'm happy to do so - we have an actual use case for that in customer code.
